Castorice - tracking

Fesaa opened the pull request.

Fesaa committed:
First walk through the kit

MISSING:
Pollux abilities
Targeting logic

Fesaa committed:
Merge branch 'Darkglade1:main' into castorice

Fesaa committed:
Migrate to new event system

Again... Need a better way to maniplulate events with the newer
characters

Fesaa committed:
Fixed lots of hooks missing @Subscribe

- Pela has an issue with her techique/talent (to fix)
- Asta gets her stacks before her energy now, so gains a little more
(fine)

Fesaa committed:
Imports

Fesaa committed:
Ensure debugs from the attack are applied

Talent: If the enemy is debuffed after Pela's attack,
So should'v been this anyway!

Fesaa committed:
Address Darklglade's comments

Fesaa committed:
More Pollux & Castorice stuff

- Revive
- Pollux skills
- Death protection

Fesaa committed:
Fix pollux being able to take too much hp

Fesaa committed:
Battle loop working, some more fixes

Lots of changes cause I was too lazy to write a full UltGoal. This is a
better way of doing them anyway!

DMG seems very low in sim..? Need to read logs

Fesaa committed:
Fix healing, kill Polux

Fesaa committed:
Fix Ruan Mei SPD boost

Fesaa committed:
Try finding better goals (didn't work lol)

Fesaa committed:
Make logs nicer to read

New lines for turns & attacks, run attack hooks after logs so order is
nicer

Fesaa committed:
Ensure hp is set before CombatStartEvent

Fesaa committed:
More Castorice/Pollux fixes

Fesaa committed:
Fix Mem Friends! Together! multiplier (E3 ups the lvl)

Fesaa committed:
Increase DMG by spamming rmc skill

Fesaa committed:
Try getting better sorted events, don't increase health by 0

Fesaa committed:
Giant Tree of Rapt Brooding

Fesaa committed:
Don't kill Pollux twice

Fesaa committed:
Use Pollux stats, but Castorice mult.

Group DmgContribution by name

Fesaa committed:
Fix poet, fix Ruan Mei skill not applying to summoned memosprites

Fesaa committed:
Include stats of (last alive version of) fallen allies

Fesaa committed:
Imports

Fesaa committed:
Formatting

Fesaa committed:
Extra guard

Fesaa committed:
Sync names

Fesaa commented on src/main/java/art/ameliah/hsr/characters/remembrance/castorice/Pollux.java:
```diff
@@ -0,0 +1,141 @@
+        });
+    }
+
+    private void BreathScorchesTheShadow(float mul) {
```
I don't think the logic here is 100% correct. I think it's given one action too many, or an option too many. I'm not sure, it's a bit hard to see in showcases.

Darkglade1 commented on src/main/java/art/ameliah/hsr/characters/remembrance/castorice/Pollux.java:
```diff
@@ -0,0 +1,141 @@
+
+    @Subscribe(priority = EventPriority.HIGHEST)
+    public void onTurnEnd(TurnEndEvent e) {
+        this.actionCounter = 0;
```
I don't think the skill's enhancement resets when the turn ends 
`Using this ability repeatedly will enhance it, lasting until the Deathwing disappears.`

Fesaa replied:
Good catch, fixed

Darkglade1 commented on src/main/java/art/ameliah/hsr/characters/remembrance/castorice/Pollux.java:
```diff
@@ -0,0 +1,141 @@
+                return;
+            }
+
+            this.die(this);
```
I'm assuming you're coding pollux to always blow up after doing 3 hp draining skills? I think whether or not that's optimal depends on the team, so it would be useful to make this configurable via goals instead of always having pollux blow up immediately.

Fesaa replied:
If pollux can isn't auto killed, doAction is called. Which triggers to goal logic, so this is possible. Currently the goals try to keep pollux alive as long as possible; https://github.com/Fesaa/Star-Rail-Battle-Simulator/blob/castorice/src/main/java/art/ameliah/hsr/characters/remembrance/castorice/PolluxSkillGoal.java

Darkglade1 commented on src/main/java/art/ameliah/hsr/characters/remembrance/castorice/Pollux.java:
```diff
@@ -0,0 +1,141 @@
+                al.hit(e, dmgMul, MultiplierStat.HP, TOUGHNESS_DAMAGE_SINGLE_UNIT);
+            });
+        }).afterAttackHook(() -> {
+            if (this.currentHp.get() > 0) {
```
`When HP is less than or equal to 25%, using this ability will actively trigger the Talent ability effect of "Wings Sweep the Ruins."` If the HP drain brings Pollux to 1 HP, Pollux will immediately blow up instead of getting another action, from the way the code is currently structured it looks like it will currently incorrectly do a "basic attack" before dying?

Fesaa replied:
Ah, so that's how it works. I wasn't sure if after draining to 1 it gets an other "turn" or not. I'll send a discord msg in a bit. i'm not sure how this fully works still

Darkglade1 commented on src/main/java/art/ameliah/hsr/characters/remembrance/castorice/Pollux.java:
```diff
@@ -0,0 +1,141 @@
+        this.actionMetric.record(MoveType.MEMOSPRITE_BASIC);
+        getBattle().addToLog(new DoMove(this, MoveType.MEMOSPRITE_BASIC));
+
+        this.RendTheRealmBeneath();
```
treating the non hp draining skill as a "basic attack" feels weird since they are both supposed to be memosprite skills (might not matter now but might matter later if a character specifically buffs "memosprite skills" or something)

Fesaa replied:
Good point, hakush has it under Memosprite skill. Will use that one

Fesaa committed:
Fix issues Darklgade found

Fesaa committed:
Ensure Pollux kills themselves, and ensure they doesn't do it twice

Fesaa committed:
3.2v3

Fesaa committed:
Reset after any ally turn

Fesaa committed:
Fix NoontideGryphon, fix Pollux turn goals...., Use real enemies

Fesaa committed:
Some stuff ig

Castorice (Sig), RMC, Ruan Mei, HouHou                       :
3,446,634.500 > 1.000 % > 1.305 %
Castorice (Sig), RMC, Ruan Mei, Gallagher                    :
3,353,869.250 > 0.973 % > 1.270 %
Castorice (Victory In a Blink), RMC, Ruan Mei, HouHou        :
3,249,949.750 > 0.969 % > 1.230 %
Castorice (Sig), Sunday, Ruan Mei, HouHou                    :
3,223,819.500 > 0.992 % > 1.221 %
Castorice (Sig), Sunday(160), Ruan Mei, HouHou               :
3,185,199.000 > 0.988 % > 1.206 %
Castorice (Victory In a Blink), RMC, Ruan Mei, Gallagher     :
3,139,005.500 > 0.985 % > 1.188 %
Castorice (Victory In a Blink), Sunday, Ruan Mei, HouHou     :
2,785,506.500 > 0.887 % > 1.055 %
Castorice (Victory In a Blink), Sunday(160), Ruan Mei, HouHou:
2,641,346.250 > 0.948 % > 1.000 %

Fesaa committed:
Goal changes & Some v3 stuff still

Fesaa committed:
v4

Fesaa committed:
v4?

Fesaa committed:
v5

Fesaa committed:
v6 ig
